src/plans/conjugate_gradient_plan.jl

```diff
@@ -507,3 +508,75 @@ function (u::SteepestDirectionUpdateRule)(
 )
     return 0.0
 end
+
+@doc raw"""
+    ConjugateGradientRestart <: DirectionUpdateRule
+
+An update rule might require a restart, that is one gradient step, if the last two gradients
+are nearly orthogonal, cf. [^HagerZhang2006], page 12 (in the pdf, 46 in Journal page numbers).
+This method acts as a _decorator_ to any existing [`DirectionUpdateRule`](@ref) `direction_update`.
+
+When obtain from the [`ConjugateGradientDescentState`](@ref)` cgs` the last
+``p_k,X_k`` and the current ``p_{k+1},X_{k+1}`` iterate and the gradient, respectively.
+
+Then a restart is performed, i.e. ``β_k = 0`` returned if
+
+```math
+    \frac{ ⟨X_{k+1}, P_{p_{k+1}\gets p_k}X_k⟩}{\lVert X_k \rVert_{p_k}} > ξ,
+```
+where ``P_{a\gets b}(⋅)`` denotes a vector transport from the tangent space at ``a`` to ``b``,
+and ``ξ`` is the `threshold`.
+
+# Constructor
+
+    PolakRibiereCoefficient(
+        direction_update::D,
+        threshold=Inf;
+        manifold = DefaultManifold(),
+        vector_transport_method::V=default_vector_transport_method(manifold),
+        a::StoreStateAction=StoreStateAction((:Iterate, :gradient, :δ)),
+    )
+
+[^HagerZhang2006]:
+    > W. W. Hager and H. Zhang, A Survey of Nonlinear Conjugate Gradient Methods
+    > Pacific Journal of Optimization, Vol. 2, 2006, pp. 35-58.
+    > url: [http://www.yokohamapublishers.jp/online2/pjov2-1.html](http://www.yokohamapublishers.jp/online2/pjov2-1.html)
+"""
+mutable struct ConjugateGradientRestart{
+    DUR<:DirectionUpdateRule,VT<:AbstractVectorTransportMethod,F
+} <: DirectionUpdateRule
+    direction_update::DUR
+    storage::StoreStateAction
+    threshold::F
+    vector_transport_method::VT
+    function ConjugateGradientRestart(
+        direction_update::D,
+        threshold=Inf;
+        manifold=DefaultManifold(),
+        vector_transport_method::V=default_vector_transport_method(manifold),
+        a::StoreStateAction=StoreStateAction((:Iterate, :gradient, :δ)),
+    ) where {D<:DirectionUpdateRule,V<:AbstractVectorTransportMethod}
+        return new{D,V,typeof(threshold)}(
+            direction_update, a, threshold, vector_transport_method
+        )
+    end
+end
+function (u::ConjugateGradientRestart)(
```

Comment:
It's a bit sad that this method has to repeat so much work that is already performed in the wrapped update rule.

Reply:
Even more, the internal one might (somehow) work without storage, then the storage here is really required.

Reply:
Sure, not every update rule has to maintain storage and do the vector transports but some do, and then the work is done twice. But it can't be generically simplified.